zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java

```diff
@@ -0,0 +1,114 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin.collector.kafka10;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.InterruptException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import zipkin.Codec;
+import zipkin.collector.Collector;
+import zipkin.collector.CollectorMetrics;
+
+import static zipkin.storage.Callback.NOOP;
+
+/** Consumes spans from Kafka messages, ignoring malformed input */
+final class KafkaCollectorWorker implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaCollectorWorker.class);
+
+  final Consumer<byte[], byte[]> kafkaConsumer;
+  final Collector collector;
+  final CollectorMetrics metrics;
+  final AtomicReference<List<TopicPartition>> assignedPartitions =
+      new AtomicReference<>(Collections.emptyList());
+
+  KafkaCollectorWorker(Properties kafkaConsumerConfig, String topic, Collector collector,
+      CollectorMetrics metrics) {
+    kafkaConsumer = new KafkaConsumer<>(kafkaConsumerConfig);
+    kafkaConsumer.subscribe(Collections.singleton(topic), new ConsumerRebalanceListener() {
+      @Override public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+        assignedPartitions.set(Collections.emptyList());
+      }
+
+      @Override public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+        assignedPartitions.set(Collections.unmodifiableList(new ArrayList<>(partitions)));
+      }
+    });
+    this.collector = collector;
+    this.metrics = metrics;
+  }
+
+  /**
+   * @return Kafka topic partitions currently assigned to this worker for processing. Returned
+   * list is not modifiable.
+   */
+  public List<TopicPartition> assignedPartitions() {
+    return assignedPartitions.get();
+  }
+
+  @Override
+  public void run() {
+    try {
+      LOG.info("Kafka consumer starting polling loop.");
+      while (true) {
+        final ConsumerRecords<byte[], byte[]> consumerRecords = kafkaConsumer.poll(1000);
+        LOG.debug("Kafka polling returned batch of {} messages.", consumerRecords.count());
+        consumerRecords.forEach((cr) -> {
```

Comment:
I think I'll need to replace the forEach() call here to maintain Java 7 compatibility.